### DISCUSSION TITLE
Can we make elements (or at least shadows) overlap other elements?

### DISCUSSION THREAD

**Discussion by andrewaclear** [tsujan/Kvantum#830] [Q&A]
I am trying to make a menu item hover have shadows behind it, but in order to do so I have to make each item far too tall because I have to account for space for the shadows as well:
![image](https://user-images.githubusercontent.com/51063733/197276783-4e3ed6ab-b2e3-42f9-a3de-da3ebb0c15ec.png)
(my Kvantum theme: menu items are too vertically spaced out)

Is there anyway I can achieve this without making space for the shadow i.e. let the shadow overlap the other menu items?
![image](https://user-images.githubusercontent.com/51063733/197277353-46e40bf3-977b-4b8f-b4e6-f17814c38924.png)
(my GTK theme: example that allows the overlapping shadows)

**Comment by tsujan** [comment]
I'm afraid I didn't understand your question. But here are some explanations:

In Kvantum, the menu shadow is the outer part of the menu rectangle itself, not a separate shadow like in Breeze. The reason is that Kvantum can be used under all DEs; so, depending on KDE for providing menu shadows wasn't an option.

Because of that, it's better not to set `menu_shadow_depth` to a too big value. Although there may be no problem with a big `menu_shadow_depth` in most apps, there are situations, where it may cause problems. I suggest `10` as the maximum value.

As a consequence, there's a limit to corner rounding in the case of menus.

This is also true for tooltips (`tooltip_shadow_depth`). In the themes that come with Kvantum, I've chosen `7` as the maximum value (it's usually `6`).

**Reply by tsujan** [reply]
You might also want to use `spread_menuitems` if you have a big `menu_shadow_depth`. See [Theme-Config](https://github.com/tsujan/Kvantum/blob/master/Kvantum/doc/Theme-Config#L495).

**Reply by andrewaclear** [reply]
I am using `spread_menuitems` so that the menu items overlap the frame horizontally, which as you said is allowing me to make my frame larger (have rounder corners and larger shadows). This is all currently working fine (the fact that I can make a shadow is perfect because I am working on this for QT apps in GNOME right now actually, which like you said does not have KDE to make the shadows for me). 

The question I was trying to ask was how to make a shadow behind **menu items** _on hover_. Right now I have achieved this by making it a part of the item's frame, but the issue is that the frame is far too big vertically (cause I am leaving space for the shadow) which makes the elements too far spaced apart. What I was asking was how can I make a menu item hover shadow that goes **around** the element, so i don't have to make them so spaced apart.

**Reply by tsujan** [reply]
> The question I was trying to ask was how to make a shadow behind menu items on hover. . Right now I have achieved this by making it a part of the item's frame,...

Oh, now I understand what you mean.

In Qt, shadows are only for windows —  menus and tooltips are also windows. Non-window widgets like menu-items, buttons, line-edits,... cannot have shadows.

Of course, you can create a "shadow effect" for any widget by using a part of its frame, but that requires a thicker frame, which results in greater vertical/horizontal spacings.

GTK is different. Apparently, its non-window widgets can also have shadows without increasing spacings.

**Reply by andrewaclear** [reply]
Ya exactly. GTK basically allows us to put shadows behind any element as it treats any "box" basically the same (elements are styled like CSS). 

So how can we get around this issue for QT apps?

**Reply by tsujan** [reply]
> So how can we get around this issue for QT apps?

There's no way — unless Qt radically changes how widgets are drawn, which is unlikely.

**Reply by andrewaclear** [reply]
There is no way to perhaps bring it to the flexibility of CSS like styling, but I think there are other ways to get around it. Oxygen is a QT theme for instance and they managed to put a glow around their menuitem hovers without shifting all the other menuitems down. How did they do that?

**Reply by tsujan** [reply]
> Oxygen is a QT theme for instance and they managed to put a glow around their menuitem hovers

Ii's simply an animation inside the menu-item's rectangle; isn't related to the kind of shadow you meant.
